File: AccessModifiers/AccessModifiers.java
/**
 * Java Access Modifiers Demonstration
 * 
 * This file demonstrates all access modifiers in Java:
 * - public
 * - private  
 * - protected
 * - default (package-private)
 */

package AccessModifiers;

// Public class - accessible from anywhere
public class AccessModifiers {
    
    // Public field - accessible from anywhere
    public String publicField = "Public field";
    
    // Private field - accessible only within this class
    private String privateField = "Private field";
    
    // Protected field - accessible within package and subclasses
    protected String protectedField = "Protected field";
    
    // Default (package-private) field - accessible within package only
    String defaultField = "Default field";
    
    // Public constructor
    public AccessModifiers() {
        System.out.println("AccessModifiers object created");
    }
    
    // Public method - accessible from anywhere
    public void publicMethod() {
        System.out.println("Public method called");
        // Can access all fields within the class
        System.out.println(publicField);
        System.out.println(privateField);
        System.out.println(protectedField);
        System.out.println(defaultField);
    }
    
    // Private method - accessible only within this class
    private void privateMethod() {
        System.out.println("Private method called");
    }
    
    // Protected method - accessible within package and subclasses
    protected void protectedMethod() {
        System.out.println("Protected method called");
    }
    
    // Default method - accessible within package only
    void defaultMethod() {
        System.out.println("Default method called");
    }
    
    // Public method that calls private method
    public void callPrivateMethod() {
        privateMethod(); // OK - within same class
    }
}

// Subclass to demonstrate protected access
class SubClass extends AccessModifiers {
    public void testAccess() {
        // Can access public, protected, and default members
        System.out.println(publicField);      // OK
        System.out.println(protectedField);   // OK
        System.out.println(defaultField);     // OK (same package)
        // System.out.println(privateField);  // ERROR - private
        
        publicMethod();      // OK
        protectedMethod();   // OK
        defaultMethod();     // OK (same package)
        // privateMethod();  // ERROR - private
    }
}

// Class in same package to demonstrate package access
class SamePackageClass {
    public void testAccess() {
        AccessModifiers obj = new AccessModifiers();
        
        // Can access public, protected, and default members
        System.out.println(obj.publicField);      // OK
        System.out.println(obj.protectedField);   // OK (same package)
        System.out.println(obj.defaultField);     // OK (same package)
        // System.out.println(obj.privateField);  // ERROR - private
        
        obj.publicMethod();      // OK
        obj.protectedMethod();   // OK (same package)
        obj.defaultMethod();     // OK (same package)
        // obj.privateMethod();  // ERROR - private
    }
}

// Demonstration class
class AccessModifiersDemo {
    public static void main(String[] args) {
        System.out.println("=== Access Modifiers Demonstration ===\n");
        
        AccessModifiers obj = new AccessModifiers();
        
        // Accessing public members
        System.out.println("Public field: " + obj.publicField);
        obj.publicMethod();
        
        // Can call private method through public method
        obj.callPrivateMethod();
        
        // Accessing protected and default (same package)
        System.out.println("\nProtected field: " + obj.protectedField);
        obj.protectedMethod();
        
        System.out.println("\nDefault field: " + obj.defaultField);
        obj.defaultMethod();
        
        // Testing with subclass
        System.out.println("\n=== Subclass Access ===");
        SubClass subObj = new SubClass();
        subObj.testAccess();
        
        // Testing with same package class
        System.out.println("\n=== Same Package Access ===");
        SamePackageClass samePackage = new SamePackageClass();
        samePackage.testAccess();
    }
}

/**
 * Real-world example: BankAccount with proper encapsulation
 */
class BankAccount {
    // Private fields - implementation details hidden
    private String accountNumber;
    private double balance;
    private String accountHolderName;
    
    // Public constructor
    public BankAccount(String accountNumber, String name, double initialBalance) {
        this.accountNumber = accountNumber;
        this.accountHolderName = name;
        this.balance = initialBalance;
    }
    
    // Public methods - public API
    public void deposit(double amount) {
        if (validateAmount(amount)) {
            balance += amount;
            logTransaction("Deposit", amount);
        }
    }
    
    public boolean withdraw(double amount) {
        if (validateAmount(amount) && balance >= amount) {
            balance -= amount;
            logTransaction("Withdrawal", amount);
            return true;
        }
        return false;
    }
    
    public double getBalance() {
        return balance;
    }
    
    public String getAccountNumber() {
        return accountNumber;
    }
    
    // Private helper methods - internal implementation
    private boolean validateAmount(double amount) {
        return amount > 0;
    }
    
    private void logTransaction(String type, double amount) {
        System.out.println(type + ": $" + amount + " | New balance: $" + balance);
    }
    
    // Protected method - for subclasses
    protected void applyInterest(double rate) {
        double interest = balance * rate;
        balance += interest;
        System.out.println("Interest applied: $" + interest);
    }
}

// Subclass demonstrating protected access
class SavingsAccount extends BankAccount {
    private double interestRate;
    
    public SavingsAccount(String accountNumber, String name, double initialBalance, double rate) {
        super(accountNumber, name, initialBalance);
        this.interestRate = rate;
    }
    
    public void addMonthlyInterest() {
        // Can call protected method from parent
        applyInterest(interestRate);
    }
}

File: AccessModifiers/AccessModifiers.cs
using System;

namespace AccessModifiersDemo
{
    /// <summary>
    /// C# Access Modifiers Demonstration
    /// 
    /// This file demonstrates all access modifiers in C#:
    /// - public
    /// - private
    /// - protected
    /// - internal
    /// - protected internal
    /// - private protected
    /// </summary>
    public class AccessModifiers
    {
        // Public field - accessible from anywhere
        public string PublicField = "Public field";
        
        // Private field - accessible only within this class
        private string privateField = "Private field";
        
        // Protected field - accessible within class and derived classes
        protected string ProtectedField = "Protected field";
        
        // Internal field - accessible within same assembly
        internal string InternalField = "Internal field";
        
        // Protected internal - accessible within assembly OR derived classes
        protected internal string ProtectedInternalField = "Protected internal field";
        
        // Private protected (C# 7.2+) - accessible within assembly AND derived classes
        private protected string PrivateProtectedField = "Private protected field";
        
        // Public constructor
        public AccessModifiers()
        {
            Console.WriteLine("AccessModifiers object created");
        }
        
        // Public method - accessible from anywhere
        public void PublicMethod()
        {
            Console.WriteLine("Public method called");
            // Can access all fields within the class
            Console.WriteLine(PublicField);
            Console.WriteLine(privateField);
            Console.WriteLine(ProtectedField);
            Console.WriteLine(InternalField);
        }
        
        // Private method - accessible only within this class
        private void PrivateMethod()
        {
            Console.WriteLine("Private method called");
        }
        
        // Protected method - accessible within class and derived classes
        protected void ProtectedMethod()
        {
            Console.WriteLine("Protected method called");
        }
        
        // Internal method - accessible within same assembly
        internal void InternalMethod()
        {
            Console.WriteLine("Internal method called");
        }
        
        // Protected internal method
        protected internal void ProtectedInternalMethod()
        {
            Console.WriteLine("Protected internal method called");
        }
        
        // Private protected method
        private protected void PrivateProtectedMethod()
        {
            Console.WriteLine("Private protected method called");
        }
        
        // Public method that calls private method
        public void CallPrivateMethod()
        {
            PrivateMethod(); // OK - within same class
        }
    }
    
    // Derived class to demonstrate protected access
    public class DerivedClass : AccessModifiers
    {
        public void TestAccess()
        {
            // Can access public, protected, internal, protected internal, and private protected
            Console.WriteLine(PublicField);                    // OK
            Console.WriteLine(ProtectedField);                 // OK - derived class
            Console.WriteLine(InternalField);                  // OK - same assembly
            Console.WriteLine(ProtectedInternalField);         // OK - derived class OR same assembly
            Console.WriteLine(PrivateProtectedField);          // OK - derived class AND same assembly
            // Console.WriteLine(privateField);                // ERROR - private
            
            PublicMethod();                   // OK
            ProtectedMethod();                // OK - derived class
            InternalMethod();                 // OK - same assembly
            ProtectedInternalMethod();        // OK - derived class OR same assembly
            PrivateProtectedMethod();         // OK - derived class AND same assembly
            // PrivateMethod();               // ERROR - private
        }
    }
    
    // Same assembly class to demonstrate internal access
    public class SameAssemblyClass
    {
        public void TestAccess()
        {
            var obj = new AccessModifiers();
            
            // Can access public, internal, and protected internal members
            Console.WriteLine(obj.PublicField);                // OK
            Console.WriteLine(obj.InternalField);              // OK - same assembly
            Console.WriteLine(obj.ProtectedInternalField);     // OK - same assembly
            // Console.WriteLine(obj.ProtectedField);          // ERROR - not derived
            // Console.WriteLine(obj.PrivateProtectedField);   // ERROR - not derived
            // Console.WriteLine(obj.privateField);            // ERROR - private
            
            obj.PublicMethod();                // OK
            obj.InternalMethod();              // OK - same assembly
            obj.ProtectedInternalMethod();     // OK - same assembly
            // obj.ProtectedMethod();          // ERROR - not derived
            // obj.PrivateProtectedMethod();   // ERROR - not derived
            // obj.PrivateMethod();            // ERROR - private
        }
    }
    
    /// <summary>
    /// Real-world example: BankAccount with proper encapsulation
    /// </summary>
    public class BankAccount
    {
        // Private fields - implementation details hidden
        private string accountNumber;
        private decimal balance;
        private string accountHolderName;
        
        // Public properties with private setters
        public string AccountNumber => accountNumber;
        public decimal Balance => balance;
        public string AccountHolderName => accountHolderName;
        
        // Public constructor
        public BankAccount(string accountNumber, string name, decimal initialBalance)
        {
            this.accountNumber = accountNumber;
            this.accountHolderName = name;
            this.balance = initialBalance;
        }
        
        // Public methods - public API
        public void Deposit(decimal amount)
        {
            if (ValidateAmount(amount))
            {
                balance += amount;
                LogTransaction("Deposit", amount);
            }
        }
        
        public bool Withdraw(decimal amount)
        {
            if (ValidateAmount(amount) && balance >= amount)
            {
                balance -= amount;
                LogTransaction("Withdrawal", amount);
                return true;
            }
            return false;
        }
        
        // Private helper methods - internal implementation
        private bool ValidateAmount(decimal amount)
        {
            return amount > 0;
        }
        
        private void LogTransaction(string type, decimal amount)
        {
            Console.WriteLine($"{type}: ${amount:F2} | New balance: ${balance:F2}");
        }
        
        // Protected method - for derived classes
        protected void ApplyInterest(decimal rate)
        {
            decimal interest = balance * rate;
            balance += interest;
            Console.WriteLine($"Interest applied: ${interest:F2}");
        }
        
        // Internal method - for same assembly
        internal void AuditAccount()
        {
            Console.WriteLine($"Auditing account: {accountNumber}");
        }
    }
    
    // Derived class demonstrating protected access
    public class SavingsAccount : BankAccount
    {
        private decimal interestRate;
        
        public SavingsAccount(string accountNumber, string name, decimal initialBalance, decimal rate)
            : base(accountNumber, name, initialBalance)
        {
            this.interestRate = rate;
        }
        
        public void AddMonthlyInterest()
        {
            // Can call protected method from parent
            ApplyInterest(interestRate);
        }
    }
    
    // Main program
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("=== Access Modifiers Demonstration ===\n");
            
            var obj = new AccessModifiers();
            
            // Accessing public members
            Console.WriteLine("Public field: " + obj.PublicField);
            obj.PublicMethod();
            
            // Can call private method through public method
            obj.CallPrivateMethod();
            
            // Accessing internal members (same assembly)
            Console.WriteLine("\nInternal field: " + obj.InternalField);
            obj.InternalMethod();
            
            // Testing with derived class
            Console.WriteLine("\n=== Derived Class Access ===");
            var derived = new DerivedClass();
            derived.TestAccess();
            
            // Testing with same assembly class
            Console.WriteLine("\n=== Same Assembly Access ===");
            var sameAssembly = new SameAssemblyClass();
            sameAssembly.TestAccess();
            
            // Real-world example
            Console.WriteLine("\n=== Bank Account Example ===");
            var account = new BankAccount("12345", "John Doe", 1000m);
            account.Deposit(500);
            account.Withdraw(200);
            Console.WriteLine($"Final balance: ${account.Balance:F2}");
            
            var savings = new SavingsAccount("67890", "Jane Smith", 5000m, 0.05m);
            savings.AddMonthlyInterest();
        }
    }
}
